Raptor: split label on Stop point and not on jpp

antoine-de opened the pull request:
Split Raptor's label by stop point and not journey pattern point anymore

Bench: 
**all:**
fr_pdl: 70s-> 60s ==> -16%
israel: 118s -> 100s  ==> -18%
idf: 54s -> 48s ==> 12%

**all but foot path**
fr_pdl: 70s-> 63s ==> -11%
israel: 118s -> 114s  ==> -3%
idf: 54s -> 49s ==> -10%

**all but foot path and best label on transfers**
fr_pdl: 70s-> 68s ==> -3%
israel: 118s -> 109s  ==> -8%
idf: 54s -> 53s ==> -3%

antoine-de committed:
Raptor: change label by jpp to label by sp

we now store label by stop point and not journey pattern point.

this thus reduce the number of stored label.

For the moment the service extension are now correctly handled

antoine-de committed:
benchmark: add catch on bench

antoine-de committed:
Raptor: add 'used_jpp' label to reconstruct path

with the service extension, it is impossible to find the used jpp, so
for the moment we get store the jpp in the labels

antoine-de committed:
Utils: get correction on idx map iterator

antoine-de committed:
Raptor: small corrections

antoine-de committed:
Raptor: lint on isochrone

antoine-de committed:
Raptor: fix bug on best_labels init

l-vincent-l commented on source/routing/raptor.cpp:
```diff
@@ -676,9 +665,10 @@ std::vector<Path> RAPTOR::compute(const type::StopArea* departure,
-    for(size_t i = 0; i <= this->count; ++i){
-        if(labels[i].dt_pt(journey_pattern_point_idx) == best_labels[journey_pattern_point_idx]){
+int RAPTOR::best_round(SpIdx sp_idx){
+    //Q: a mon avis on peut virer cette methode
```
In english

l-vincent-l commented on source/routing/raptor.h:
```diff
@@ -62,43 +62,35 @@ struct RoutingState {
-    boost::dynamic_bitset<> valid_journey_pattern_points;
     dataRAPTOR::JppsFromSp jpps_from_sp;
-    ///L'ordre du premier j: public AbstractRouterourney_pattern point de la journey_pattern
+    ///L'ordre du premier journey_pattern point de la journey_pattern
```
Translate

l-vincent-l commented on source/routing/raptor_path.cpp:
```diff
@@ -45,26 +45,38 @@ makePathes(const std::vector<std::pair<SpIdx, navitia::time_duration> > &departu
-    if(ls.pt_is_initialized(jpp_idx)) {
-        const auto& dt_pt = ls.dt_pt(jpp_idx);
+
+    if(ls.pt_is_initialized(sp_idx)) { //TODO donc ca c'est a enlever,checke au dessus
```
In english please!

l-vincent-l commented on source/routing/raptor_path.cpp:
```diff
@@ -45,26 +45,38 @@ makePathes(const std::vector<std::pair<SpIdx, navitia::time_duration> > &departu
+    const auto jpp_idx = get_used_jpp(count, sp_idx, raptor);
+
+    if (! jpp_idx.is_valid()) {
+        throw "c'est la merde";
```
This is bullshit!

l-vincent-l commented on source/routing/raptor_path.cpp:
```diff
@@ -286,12 +298,12 @@ void patch_datetimes(Path &path){
+//                const type::AccessibiliteParams & accessibilite_params,
+//                const RAPTOR &raptor_, bool disruption_active) {
+//    return makePath(destination_idx, countb, disruption_active, false, accessibilite_params, raptor_);
+//}
```
You may delete it.

l-vincent-l commented on source/routing/raptor_path_defs.h:
```diff
@@ -89,21 +112,25 @@ handle_st(const type::StopTime* st, DateTime& workingDate, bool clockwise, const
-
+/**
+ *TODO remonte le vj + extension
+ */
```
In english please!

l-vincent-l commented on source/routing/raptor_solutions.cpp:
```diff
@@ -137,75 +135,75 @@ get_pareto_front(bool clockwise, const std::vector<std::pair<SpIdx, navitia::tim
+            SpIdx final_sp_idx;
+            std::tie(final_sp_idx, s.upper_bound) = get_final_spidx_and_date(round, best_sp, clockwise,
                                                                              disruption_active, accessibilite_params, raptor);
+            //TODO l'heure de depart (upper bound) n'est pas bon pour le moment, voir is il ya moyen de mieux le calculer
```
In english please!

l-vincent-l commented on source/routing/raptor_solutions.cpp:
```diff
@@ -231,74 +229,75 @@ get_walking_solutions(bool clockwise, const std::vector<std::pair<SpIdx, navitia
+                    best.total_arrival;
+
+
+            //Si je gagne 5 minutes de marche a pied, je suis pret à perdre jusqu'à 10 minutes.
```
In english please!

l-vincent-l commented on source/routing/raptor_solutions.cpp:
```diff
@@ -320,59 +319,56 @@ get_walking_solutions(bool clockwise, const std::vector<std::pair<SpIdx, navitia
     }
 };
 
 // Reparcours l’itinéraire rapidement pour avoir le JPP et la date de départ (si on cherchait l’arrivée au plus tôt)
```
In english please!

antoine-de committed:
Raptor: lint

antoine-de committed:
Raptor: add filter on journeys

we nos post filter journeys with zone ODT transfers

antoine-de committed:
Utils: update to have 'at' on IdxMap

antoine-de committed:
Raptor: lint

antoine-de committed:
Raptor: change IdxMap for more typecheck

antoine-de committed:
Raptor: add typecheck on IdxMap constructor

antoine-de commented:
Jenkins, test this please

antoine-de committed:
TimeTable: change init

that was causing a strange bug because the data were sorted after the
build_raptor...

TeXitoi commented on source/routing/dataraptor.cpp:
```diff
@@ -37,8 +37,8 @@ www.navitia.io
-    forward_connections.assign(data.stop_points.size());
-    backward_connections.assign(data.stop_points.size());
+    forward_connections.assign(data.stop_points);
+    backward_connections.assign(data.stop_points);
```
Great!

antoine-de committed:
Raptor: remove catch on benchmark

it's better to fail

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -82,23 +77,19 @@ bool RAPTOR::apply_vj_extension(const Visitor& v, const bool global_pruning,
-            auto& best_jpp = best_jpp_by_sp[SpIdx(*jpp->stop_point)];
-            if(!best_jpp.is_valid() || v.comp(workingDt, working_labels.dt_pt(best_jpp))) {
-                best_jpp = jpp_idx;
-                result = true;
```
I think you should not remove this line. Please do TTD on this possible bug ;)

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -113,37 +104,37 @@ bool RAPTOR::apply_vj_extension(const Visitor& v, const bool global_pruning,
+            }
+
+            // we mark the jpp order
+            for(const auto& jpp: jpps_from_sp[destination_sp_idx]) {
```
You should try to put that after every footpath

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -151,13 +142,16 @@ bool RAPTOR::foot_path(const Visitor & v) {
         }
     }
+
+    //TODO try to loop on all touched stop point here
```
yes, just that

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -179,29 +171,28 @@ void RAPTOR::init(Solutions departs,
-                }
-            }
-        }
+        if (! sp->accessible(required_properties)) { continue; }
```
This test is already done in jpps_from_sp

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -587,20 +580,16 @@ void RAPTOR::raptor_loop(Visitor visitor, const type::AccessibiliteParams & acce
-                                auto& best_jpp = best_jpp_by_sp[jpp.sp_idx];
-                                if(!best_jpp.is_valid() || visitor.comp(workingDt, working_labels.dt_pt(best_jpp))) {
-                                    best_jpp = jpp.idx;
-                                    end_algorithm = false;
```
Same with this one, and TDD will be easier

TeXitoi commented on source/routing/raptor.h:
```diff
@@ -62,43 +62,35 @@ struct RoutingState {
 
     NextStopTime next_st;
 
     ///Contient les heures d'arrivées, de départ, ainsi que la façon dont on est arrivé à chaque journey_pattern point à chaque tour
```
Maybe you can translate a bit more around here.

TeXitoi commented on source/routing/raptor.h:
```diff
@@ -62,43 +62,35 @@ struct RoutingState {
+    ///Order of the first journey_pattern point of each journey_pattern
     IdxMap<type::JourneyPattern, int> Q;
 
     //Constructeur
```
Just remove this incredible comment ;)

TeXitoi commented on source/routing/raptor.h:
```diff
@@ -211,7 +202,7 @@ struct RAPTOR
 
     /// Retourne à quel tour on a trouvé la meilleure solution pour ce journey_patternpoint
     /// Retourne -1 s'il n'existe pas de meilleure solution
```
Please translate

TeXitoi commented on source/routing/raptor_api.cpp:
```diff
@@ -507,50 +507,42 @@ static void add_isochrone_response(RAPTOR& raptor,
+                    sp_idx, !clockwise, disruption_active, accessibilite_params, raptor);
+            DateTime initial_dt = sp_and_date.second;
+
+            int duration = ::abs(best_lbl - init_dt); //Q: I don't see the point of the get_final_spidx_and_date call
```
Maybe remove this comment?

antoine-de replied:
We need to understand this bit of code before merging

TeXitoi commented on source/routing/raptor_api.cpp:
```diff
@@ -507,50 +507,42 @@ static void add_isochrone_response(RAPTOR& raptor,
+            int duration = ::abs(best_lbl - init_dt); //Q: I don't see the point of the get_final_spidx_and_date call
+
+            if(duration > max_duration) {
+                continue;
```
I love that, so more readable!

TeXitoi commented on source/routing/raptor_api.cpp:
```diff
@@ -507,50 +507,42 @@ static void add_isochrone_response(RAPTOR& raptor,
+            pb_journey->set_departure_date_time(str_departure);
+            pb_journey->set_requested_date_time(str_requested);
+            pb_journey->set_duration(duration);
+            pb_journey->set_nb_transfers(round); //Q: round there ? if round = 1, there are 0 transfer no ?
```
We have to answer that too

TeXitoi commented on source/routing/raptor_api.cpp:
```diff
@@ -507,50 +507,42 @@ static void add_isochrone_response(RAPTOR& raptor,
+                                  raptor.data, pb_journey->mutable_destination(), 0, now, action_period, show_codes);
+            else
+                fill_pb_placemark(sp,
+                                  raptor.data, pb_journey->mutable_destination(), 0, now, action_period, show_codes);
```
long lines

TeXitoi commented on source/routing/raptor_path.cpp:
```diff
@@ -125,14 +133,14 @@ get_current_stidx_gap(size_t count,
+        void connection(const type::StopPoint* departure, const type::StopPoint* destination,
                     boost::posix_time::ptime dep_time, boost::posix_time::ptime arr_time,
-                    type::StopPointConnection* stop_point_connection) {
+                    const type::StopPointConnection* stop_point_connection) {
```
Great!

TeXitoi commented on source/routing/raptor_path.h:
```diff
@@ -94,7 +86,7 @@ namespace navitia { namespace routing {
         }
 
-        void final_step(JppIdx /*current_jpp*/, size_t /*count*/, const std::vector<Labels> &/*labels*/) {
+        void final_step(SpIdx /*current_jpp*/, size_t /*count*/, const std::vector<Labels> &/*labels*/) {
```
update the variable names (or just remove them)

TeXitoi commented on source/routing/raptor_path_defs.h:
```diff
@@ -6,26 +6,48 @@
+ * Returns the jpp used to mark the stop point in the label
+ *
+ * Note: this should not be necessary, since we should be able to recompute this used jpp
+ * (by iterating on the jpp of the label.boarding_jpp's journey pattern), but because of service extension we cannot do that
```
long line

TeXitoi commented on source/routing/raptor_path_defs.h:
```diff
@@ -89,21 +111,27 @@ handle_st(const type::StopTime* st, DateTime& workingDate, bool clockwise, const
                                                               raptor_, disruption_active);
     while(boarding_jpp_idx != current_jpp_idx) {
         // There is a side effect on workingDate caused by workingDate
+        //TODO remove the side effect
```
You want to do it or it's for the next one?

antoine-de replied:
yup for the next one :)

TeXitoi commented on source/routing/raptor_solution_filter.cpp:
```diff
@@ -0,0 +1,64 @@
+/* Copyright © 2001-2014, Canal TP and/or its affiliates. All rights reserved.
```
update the date ;)

antoine-de committed:
Raptor: corrections after review

TeXitoi commented on source/routing/raptor_solutions.cpp:
```diff
@@ -231,74 +229,75 @@ get_walking_solutions(bool clockwise, const std::vector<std::pair<SpIdx, navitia
+
+            SpIdx sp_idx = spid_dist.first;
+
+            // We only want solution ending by a vehicle journey or a stay_in
```
This comment seems totally nonsense for me. Remove it? rephrase it?

l-vincent-l replied:
This is oppose to journeys ending by a transfer

TeXitoi commented on source/time_tables/tests/departure_boards_test.cpp:
```diff
@@ -432,3 +433,16 @@ BOOST_FIXTURE_TEST_CASE(test_calendar_start_time_period_before, small_cal_fixtur
+    navitia::routing::RAPTOR raptor(*(b.data));
+    navitia::type::PT_Data& d = *b.data->pt_data;
+
+    std::cout << "== " << d.stop_areas_map["stop1"] << std::endl;
```
I think you want to remove this cout

TeXitoi commented:
Great patch!

antoine-de committed:
Raptor: correction after review

l-vincent-l commented on source/routing/raptor_path.cpp:
```diff
@@ -45,26 +45,34 @@ makePathes(const std::vector<std::pair<SpIdx, navitia::time_duration> > &departu
         const auto date = DateTimeUtils::date(dt_pt);
         const auto hour = DateTimeUtils::hour(dt_pt);
         const type::JourneyPatternPoint* jpp = raptor.get_jpp(jpp_idx);
+
```
I love that! this is much more readable.

l-vincent-l commented on source/routing/raptor_path.h:
```diff
@@ -94,7 +86,7 @@ namespace navitia { namespace routing {
         }
 
-        void final_step(JppIdx /*current_jpp*/, size_t /*count*/, const std::vector<Labels> &/*labels*/) {
+        void final_step(SpIdx /*current_sp*/, size_t /*count*/, const std::vector<Labels> &/*labels*/) {
```
const no?

antoine-de committed:
Raptor: split best label in 2

we now have 2 best labels, one to arrive by pt on the stop point, and
one for transfers.
It helps prune the search.

add some misc corrections after review too

antoine-de committed:
Raptor: small corrections

antoine-de committed:
Raptor: move foot path result for coherence

antoine-de committed:
Raptor: add test on raptor stop

antoine-de committed:
Bench: add callgrind instruction
